Feature: Instruction Schema Validation of well known blueprints

0xOmarA opened the pull request.

iamyulong commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+                instruction_index,
+                invocation.global_address(),
+            ));
+        };
```
Minor: just to note that all addresses are entity-type-checked.

0xOmarA replied:
Gotcha, so you're saying that unwrapping here is safe to do?

iamyulong commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+        };
+
+    let blueprint_schema = match invocation {
+        Invocation::Function(package_address @ PACKAGE_PACKAGE, ref blueprint, _) => {
```
`@` looks fancy!

0xOmarA replied:
Can't say no to using fancy rust features :)

dhedey commented on transaction/examples/account/new.rtm:
```diff
@@ -1,3 +1,4 @@
 CREATE_ACCOUNT_ADVANCED
-    Map<String, Tuple>();                  # Authority Rules
+    Enum<0u8>();
```
In all of these `.rtm` files, can we instead use the string based values from `enums.rs` so people know what the type of this is?

0xOmarA replied:
Sounds good, will update that.

dhedey commented on transaction/src/manifest/e2e.rs:
```diff
@@ -240,25 +240,25 @@ CALL_FUNCTION
                 r##"
 CALL_METHOD
-    Address("${component_address}")
+    Address("component_sim1cqvgx33089ukm2pl97pv4max0x40ruvfy4lt60yvya744cvemygpmu")
```
Why have we moved away from using `component_address` / `account_address` / `package_address` here? :)

0xOmarA replied:
My bad, switching back.

dhedey commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+    Ok(())
+}
+
+macro_rules! define_lazy_package_definitions {
```
Out of interest, why don't we use lazy static here?

It would be helpful to have this statically defined in the node :)

0xOmarA replied:
It can definitely use lazy static here. My thought when I did this was that we would only be using it here, thus having the schemas as globals didn't make a lot of sesne. However, as you mentioned in your other comment the node wants to use this, and so does the toolkit. So, I will use lazy static here.

dhedey commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+
+// `usize` here is the index of the instruction has has an error.
+#[derive(Clone, Debug)]
+pub enum ValidationError {
```
This is exposed as `radix_engine::utils::ValidationError` which doesn't feel like quite the right description.

Perhaps `ManifestSchemaValidationError`? Or else put this in `radix_engine::utils::schema::ValidationError` instead?

dhedey reviewed (approved):
Super nice!

Big thanks for fixing up all the manifests - it's a shame how much they managed to diverge when we didn't have these tests in place!

So happy to finally have all this come together! And I'm actually happily surprised how simple the native blueprint call validator code looks.

Have shared a few minor suggestions for some tweaks.

Relatedly, some minor thoughts on extensions - what are your thoughts on:
* We can re-use the blueprint lookup for formatting with RustLike syntax for a nicer instruction display (to eg show users what their access rules are)
* We can provide some kind of support for listing the non-native schemas we could use alongside the resolution result, and/or providing these non-native schemas directly.

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+    blueprint: &str,
+    instruction_index: usize,
+) -> Result<&'p BlueprintSchema, ValidationError> {
+    package_definition.schema.blueprints.get(blueprint).map_or(
```
Nice, did not know about that! Will start using that.

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+    };
+}
+
+define_lazy_package_definitions! {
```
Yup, will convert this to be lazy static and will make sure that they're public so the node and toolkit can use them.

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+
+// `usize` here is the index of the instruction has has an error.
+#[derive(Clone, Debug)]
+pub enum ValidationError {
```
Sounds good! I've went ahead with names similar to the ones you have in your schema implementation. 

* `LocatedInstructionSchemaValidationError` for the error that includes the `cause` and the `instruction_index`.
* `InstructionSchemaValidationError` for the error enum

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+        }
+    }
+
+    fn global_address(&self) -> GlobalAddress {
```
I've replaced that with an `entity_type` method that returns the `EntityType` of the object being invoked. How does that sound?

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+                .get(NON_FUNGIBLE_VAULT_BLUEPRINT),
+
+            EntityType::GlobalGenericComponent
+            | EntityType::InternalGenericComponent
```
I mainly have support for it in anticipation for Yulong's item from the "To Discuss" document that recommends that we add something along the lines of `Expression("LAST_OWNED")`.

0xOmarA commented on radix-engine/src/utils/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,460 @@
+                ),
+                args,
+            ),
+            _ => continue,
```
Yup, I have added that and have added support for direct access methods.

0xOmarA commented on radix-engine-tests/tests/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,243 @@
+    ComponentAddress::virtual_account_from_public_key(&private_key2().public_key())
+}
+
+fn apply_replacements(string: &str) -> String {
```
Sounds good, will use the ones from `e2e` here to avoid duplication.

0xOmarA commented on transaction/examples/call/call_method.rtm:
```diff
@@ -6,15 +6,16 @@ CALL_METHOD
-    Decimal("1");
+    "set_royalty"
+    "my_method"
+    Enum<0u8>();
```
Yup, sounds good.

0xOmarA commented on transaction/examples/access_rule/access_rule.rtm:
```diff
@@ -1,6 +1,5 @@
-    "auth"
-    Enum<1u8>(Enum<AccessRule::AllowAll>())
-    Enum<0u8>();
+    "hello"
```
Sure thing, will add that.

0xOmarA commented on radix-engine-tests/tests/native_blueprint_call_validator.rs:
```diff
@@ -0,0 +1,243 @@
+    let validation_result = validate_call_arguments_to_native_components(&manifest.instructions);
+
+    // Assert
+    assert!(validation_result.is_ok())
```
Sounds good, We now use `.except(message)` here.

0xOmarA commented:
Thanks @iamyulong and @dhedey for the review! I've addressed and comments and updated the PR.

If we later on decide that we wish to add this to the `NotarizedTransactionValidator` it should be a very simple addition to add. For now, this will be used in the simulator and the toolkit.

> We can re-use the blueprint lookup for formatting with RustLike syntax for a nicer instruction display (to eg show users what their access rules are)

Certainly! I think we will first need to do your second point of allowing the caller to provide the schema of non-native components and packages. With that, we can format all instructions in a nice rust-like way.

> We can provide some kind of support for listing the non-native schemas we could use alongside the resolution result, and/or providing these non-native schemas directly.

This is definitely doable, in fact I imagine that it would be easy to do. The validation function (whatever the name we pick ends up being) could take an additional argument which is a `BTreeMap<NodeId, &'p PackageDefinition>` we can then use that package definition when we're trying to lookup the associated schema depending on whether it's a method or function call.
